File: 13. Rekursi Lanjut/C. Permutasi Zig-Zag/c_permutasi_zig_zag.pas
program c_permutasi_zigzag;

var
N, i : integer;
isVisited: array[1..17] of boolean;
save: array[1..17] of integer;

procedure draw(depth : integer);
var
isZigZag, h, l : boolean;
j, k : integer;
begin
	if (depth > N) then begin
		isZigZag := true;
		if (N >= 3) then begin
			for j := 2 to N-1 do begin
				h := (save[j] > save[j-1]) and (save[j] > save[j + 1]);
				l := (save[j] < save[j - 1]) and (save[j] < save[j + 1]);
				if not(h or l) then isZigZag := false;
			end;
		end;
		if (N < 3) or (isZigZag = true) then begin
			for j := 1 to N do write(save[j]);
			writeln;
		end;
	end
	else begin
		for k := 1 to N do begin
			if (not isVisited[k]) then begin
				isVisited[k] := true;
				save[depth] := k;
				draw(depth + 1);
				isVisited[k] := false;
			end;
		end;
	end;
end;

begin
	readln(N);
	for i := 1 to 17 do isVisited[i] := false;
		
    draw(1);
end.

File: 13. Rekursi Lanjut/C. Permutasi Zig-Zag/c_permutasi_zig_zag.rs
fn write(n: i32, depth: i32, save: &mut Vec<i32>, is_visited: &mut Vec<bool>) {
    if depth > n {
        let mut is_zig_zag: bool = true;
        if n >= 3 {
            for i in 2..n {
                let i: usize = i as usize;
                let h = save[i] > save[i - 1] && save[i] > save[i + 1];
                let l = save[i] < save[i - 1] && save[i] < save[i + 1];
                if !(h || l) {
                    is_zig_zag = false;
                    break;
                }
            }
        }
        if n < 3 || is_zig_zag {
            for i in 1..n + 1 {
                print!("{}", save[i as usize]);
            }
            println!();
        }
    } else {
        for i in 1..n + 1 {
            if !is_visited[i as usize] {
                is_visited[i as usize] = true;
                save[depth as usize] = i;
                write(n, depth + 1, save, is_visited);
                is_visited[i as usize] = false;
            }
        }
    }
}

fn main() {
    let mut reader = String::new();
    std::io::stdin().read_line(&mut reader).expect("");
    let n: i32 = reader.trim().parse().expect("");
    let mut save: Vec<i32> = vec![0; 17];
    let mut is_visited: Vec<bool> = vec![false; 17];
    write(n, 1, &mut save, &mut is_visited);
}
